ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java

```diff
@@ -0,0 +1,552 @@
+  }
+
+  private Vertx vertx;
+  private Client client;
+
+  @Before
+  public void setUp() {
+    vertx = Vertx.vertx();
+    client = createClient();
+  }
+
+  @After
+  public void tearDown() {
+    if (client != null) {
+      client.close();
+    }
+    if (vertx != null) {
+      vertx.close();
+    }
+    REST_APP.getServiceContext().close();
+  }
+
+  @Test
+  public void shouldStreamPushQueryAsync() throws Exception {
+    // When
+    final StreamedQueryResult streamedQueryResult = client.streamQuery(PUSH_QUERY).get();
+
+    // Then
+    assertThat(streamedQueryResult.columnNames(), is(PAGE_VIEW_COLUMN_NAMES));
+    assertThat(streamedQueryResult.columnTypes(), is(PAGE_VIEW_COLUMN_TYPES));
+    assertThat(streamedQueryResult.queryID(), is(notNullValue()));
+
+    shouldReceivePageViewRows(streamedQueryResult, false);
+
+    assertThat(streamedQueryResult.isComplete(), is(false));
+  }
+
+  @Test
+  public void shouldStreamPushQuerySync() throws Exception {
+    // When
+    final StreamedQueryResult streamedQueryResult = client.streamQuery(PUSH_QUERY).get();
+
+    // Then
+    assertThat(streamedQueryResult.columnNames(), is(PAGE_VIEW_COLUMN_NAMES));
+    assertThat(streamedQueryResult.columnTypes(), is(PAGE_VIEW_COLUMN_TYPES));
+    assertThat(streamedQueryResult.queryID(), is(notNullValue()));
+
+    for (int i = 0; i < PAGE_VIEW_NUM_ROWS; i++) {
+      final Row row = streamedQueryResult.poll();
+      verifyPageViewRowWithIndex(row, i);
+    }
+
+    assertThat(streamedQueryResult.isComplete(), is(false));
+  }
+
+  @Test
+  public void shouldStreamPullQueryAsync() throws Exception {
+    // When
+    final StreamedQueryResult streamedQueryResult = client.streamQuery(PULL_QUERY).get();
+
+    // Then
+    assertThat(streamedQueryResult.columnNames(), is(PULL_QUERY_COLUMN_NAMES));
+    assertThat(streamedQueryResult.columnTypes(), is(PULL_QUERY_COLUMN_TYPES));
+    assertThat(streamedQueryResult.queryID(), is(nullValue()));
+
+    shouldReceivePullQueryRow(streamedQueryResult);
+
+    assertThatEventually(streamedQueryResult::isComplete, is(true));
+  }
+
+  @Test
+  public void shouldStreamPullQuerySync() throws Exception {
+    // When
+    final StreamedQueryResult streamedQueryResult = client.streamQuery(PULL_QUERY).get();
+
+    // Then
+    assertThat(streamedQueryResult.columnNames(), is(PULL_QUERY_COLUMN_NAMES));
+    assertThat(streamedQueryResult.columnTypes(), is(PULL_QUERY_COLUMN_TYPES));
+    assertThat(streamedQueryResult.queryID(), is(nullValue()));
+
+    final Row row = streamedQueryResult.poll();
+    verifyPullQueryRow(row);
+    assertThat(streamedQueryResult.poll(), is(nullValue()));
+
+    assertThatEventually(streamedQueryResult::isComplete, is(true));
+  }
+
+  @Test
+  public void shouldStreamPushQueryWithLimitAsync() throws Exception {
+    // When
+    final StreamedQueryResult streamedQueryResult = client.streamQuery(PUSH_QUERY_WITH_LIMIT).get();
```

Comment:
I think the limit here is just the number of rows - would be better imho to test with limit < number of rows to make sure the limit really works.

Reply:
Seems more like server-side functionality testing than client functionality testing but there's no harm in testing it here too so I've made the update.